Investigate intermittent test failures in `PushManifestListTestCase`

```
=================================== FAILURES ===================================
____________ PushManifestListTestCase.test_push_manifest_list_v2s2 _____________

self = <pulp_container.tests.functional.api.test_push_content.PushManifestListTestCase testMethod=test_push_manifest_list_v2s2>

    def test_push_manifest_list_v2s2(self):
        """Push the created manifest list in the v2s2 format."""
        self.registry.login(
            "-u", self.user_admin["username"], "-p", self.user_admin["password"], self.registry_name
        )
>       self.registry.manifest_push(
            self.v2s2_tag,
            self.v2s2_image_path,
            "--all",
            "--format",
            "v2s2",
        )

usr/local/lib/python3.8/site-packages/pulp_container/tests/functional/api/test_push_content.py:432: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
usr/local/lib/python3.8/site-packages/pulp_smash/cli.py:975: in <lambda>
    self.manifest_push = lambda *args: self._dispatch_command(
usr/local/lib/python3.8/site-packages/pulp_smash/cli.py:1036: in _dispatch_command
    result = self._client.run(cmd)
usr/local/lib/python3.8/site-packages/pulp_smash/cli.py:322: in run
    return self.response_handler(completed_process)
usr/local/lib/python3.8/site-packages/pulp_smash/cli.py:54: in code_handler
    completed_proc.check_returncode()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = CompletedProcess(args=('podman', 'manifest', 'push', 'manifest_list', 'pulp.example.com:443/foo:manifest_list', '--all...: initiating layer upload to /v2/foo/blobs/uploads/ in pulp.example.com:443: name invalid: Invalid repository name.\n')

    def check_returncode(self):
        """Raise an exception if ``returncode`` is non-zero.
    
        Raise :class:`pulp_smash.exceptions.CalledProcessError` if
        ``returncode`` is non-zero.
    
        Why not raise ``subprocess.CalledProcessError``? Because stdout and
        stderr are not included when str() is called on a CalledProcessError
        object. A typical message is::
    
            "Command '('ls', 'foo')' returned non-zero exit status 2"
    
        This information is valuable. One could still make
        ``subprocess.CalledProcessError`` work by overloading ``args``:
    
        >>> if isinstance(args, (str, bytes)):
        ...     custom_args = (args, stdout, stderr)
        ... else:
        ...     custom_args = tuple(args) + (stdout, stderr)
        >>> subprocess.CalledProcessError(args, returncode)
    
        But this seems like a hack.
    
        In addition, it's generally good for an application to raise expected
        exceptions from its own namespace, so as to better abstract away
        dependencies.
        """
        if self.returncode != 0:
>           raise exceptions.CalledProcessError(
                self.args, self.returncode, self.stdout, self.stderr
            )
E           pulp_smash.exceptions.CalledProcessError: Command ('podman', 'manifest', 'push', 'manifest_list', 'pulp.example.com:443/foo:manifest_list', '--all', '--format', 'v2s2', '--tls-verify=true') returned non-zero exit status 125.
E           
E           stdout: 
E           
E           stderr: Getting image list signatures
E           Copying 3 of 3 images in list
E           Copying image sha256:d8fbbbf3fec1857c32c110292a9decf9744f9f97d7247019ae4776c241395221 (1/3)
E           Getting image source signatures
E           Copying blob sha256:463b9bfafbaf88a2dee0aa77fd293a7cb91558af9664bc23d4ed00830463615d
E           Copying blob sha256:d7ab49ce0b755a907e914d148cc8c1f437f7688048aea18a15d965f60c5d0a43
E           Error: copying image 1/3 from manifest list: writing blob: initiating layer upload to /v2/foo/blobs/uploads/ in pulp.example.com:443: name invalid: Invalid repository name.
```

The issue persists after introducing the above change:
```
            )
E           pulp_smash.exceptions.CalledProcessError: Command ('podman', 'manifest', 'push', 'manifest_list_oci', 'pulp.example.com:443/foo_oci:manifest_list_oci', '--all', '--format', 'oci', '--tls-verify=true') returned non-zero exit status 125.
E           
E           stdout: 
E           
E           stderr: Getting image list signatures
E           Copying 3 of 3 images in list
E           Copying image sha256:d8fbbbf3fec1857c32c110292a9decf9744f9f97d7247019ae4776c241395221 (1/3)
E           Getting image source signatures
E           Copying blob sha256:463b9bfafbaf88a2dee0aa77fd293a7cb91558af9664bc23d4ed00830463615d
E           Copying blob sha256:d7ab49ce0b755a907e914d148cc8c1f437f7688048aea18a15d965f60c5d0a43
E           Error: copying image 1/3 from manifest list: writing blob: initiating layer upload to /v2/foo_oci/blobs/uploads/ in pulp.example.com:443: name invalid: Invalid repository name.
```

--- a/CHANGES/1275.misc
+++ b/CHANGES/1275.misc
@@ -1,0 +1,2 @@
+Ensured repositories are correctly retrieved or initialized when handling client's parallel blob
+uploading.

--- a/CHANGES/489.misc
+++ b/CHANGES/489.misc
@@ -1,0 +1,1 @@
+Refactored ``get_dr_push()`` to be more readable.

--- a/pulp_container/app/registry_api.py
+++ b/pulp_container/app/registry_api.py
@@ -25,14 +25,13 @@ from django.conf import settings
 from pulpcore.plugin.models import Artifact, ContentArtifact, UploadChunk
 from pulpcore.plugin.files import PulpTemporaryUploadedFile
 from pulpcore.plugin.tasking import add_and_remove, dispatch
-from pulpcore.plugin.util import get_objects_for_user
+from pulpcore.plugin.util import get_objects_for_user, get_url
 from rest_framework.exceptions import (
     AuthenticationFailed,
     NotAuthenticated,
     PermissionDenied,
     ParseError,
     Throttled,
-    ValidationError,
 )
 from rest_framework.generics import ListAPIView
 from rest_framework.pagination import BasePagination
@@ -291,32 +290,7 @@ class ContainerRegistryApiMixin:
             distribution = models.ContainerDistribution.objects.get(base_path=path)
         except models.ContainerDistribution.DoesNotExist:
             if create:
-                try:
-                    with transaction.atomic():
-                        repo_serializer = serializers.ContainerPushRepositorySerializer(
-                            data={"name": path}, context={"request": request}
-                        )
-                        repo_serializer.is_valid(raise_exception=True)
-                        repository = repo_serializer.create(repo_serializer.validated_data)
-                        repo_href = serializers.ContainerPushRepositorySerializer(
-                            repository, context={"request": request}
-                        ).data["pulp_href"]
-
-                        dist_serializer = serializers.ContainerDistributionSerializer(
-                            data={"base_path": path, "name": path, "repository": repo_href}
-                        )
-                        dist_serializer.is_valid(raise_exception=True)
-                        distribution = dist_serializer.create(dist_serializer.validated_data)
-                except ValidationError:
-                    raise RepositoryInvalid(name=path)
-                except IntegrityError:
-                    # Seems like another process created our stuff already. Retry fetching it.
-                    distribution = models.ContainerDistribution.objects.get(base_path=path)
-                    repository = distribution.repository
-                    if repository:
-                        repository = repository.cast()
-                        if not repository.PUSH_ENABLED:
-                            raise RepositoryInvalid(name=path, message="Repository is read-only.")
+                distribution, repository = self.create_dr(path, request)
             else:
                 raise RepositoryNotFound(name=path)
         else:
@@ -326,27 +300,31 @@ class ContainerRegistryApiMixin:
                 if not repository.PUSH_ENABLED:
                     raise RepositoryInvalid(name=path, message="Repository is read-only.")
             elif create:
-                try:
-                    with transaction.atomic():
-                        repo_serializer = serializers.ContainerPushRepositorySerializer(
-                            data={"name": path}, context={"request": request}
-                        )
-                        repo_serializer.is_valid(raise_exception=True)
-                        repository = repo_serializer.create(repo_serializer.validated_data)
-                        distribution.repository = repository
-                        distribution.save()
-                except IntegrityError:
-                    # Seems like another process created our stuff already. Retry fetching it.
-                    distribution = models.ContainerDistribution.objects.get(pk=distribution.pk)
-                    repository = distribution.repository
-                    if repository:
-                        repository = repository.cast()
-                        if not repository.PUSH_ENABLED:
-                            raise RepositoryInvalid(name=path, message="Repository is read-only.")
-                    else:
-                        raise RepositoryNotFound(name=path)
+                with transaction.atomic():
+                    repository = serializers.ContainerPushRepositorySerializer.get_or_create(
+                        {"name": path}
+                    )
+                    distribution.repository = repository
+                    distribution.save()
             else:
                 raise RepositoryNotFound(name=path)
+        return distribution, repository
+
+    def create_dr(self, path, request):
+        with transaction.atomic():
+            repository = serializers.ContainerPushRepositorySerializer.get_or_create({"name": path})
+            distribution = serializers.ContainerDistributionSerializer.get_or_create(
+                {"base_path": path, "name": path}, {"repository": get_url(repository)}
+            )
+
+            if distribution.repository:
+                dist_repository = distribution.repository.cast()
+                if not dist_repository.PUSH_ENABLED or repository != dist_repository:
+                    raise RepositoryInvalid(name=path, message="Repository is read-only.")
+            else:
+                distribution.repository = repository
+                distribution.save()
+
         return distribution, repository
 
 

--- a/pulp_container/app/serializers.py
+++ b/pulp_container/app/serializers.py
@@ -196,7 +196,7 @@ class ContainerRepositorySerializer(RepositorySerializer):
         model = models.ContainerRepository
 
 
-class ContainerPushRepositorySerializer(RepositorySerializer):
+class ContainerPushRepositorySerializer(RepositorySerializer, GetOrCreateSerializerMixin):
     """
     Serializer for Container Push Repositories.
     """
@@ -277,7 +277,7 @@ class ContainerRemoteSerializer(RemoteSerializer):
         model = models.ContainerRemote
 
 
-class ContainerDistributionSerializer(DistributionSerializer):
+class ContainerDistributionSerializer(DistributionSerializer, GetOrCreateSerializerMixin):
     """
     A serializer for ContainerDistribution.
     """

--- a/requirements.txt
+++ b/requirements.txt
@@ -1,5 +1,5 @@
 ecdsa>=0.14,<=0.18.0
 jsonschema>=4.4,<4.20
-pulpcore>=3.40.1,<3.55
+pulpcore>=3.40.3,<3.55
 pyjwkest>=1.4,<=1.4.2
 pyjwt[crypto]>=2.4,<2.9

